torchrl/modules/functional_modules.py

```diff
@@ -0,0 +1,227 @@
+                    f"Got in_dim={in_dim} for an input but the input is of type "
+                    f"{type(arg)}. We cannot vmap over non-Tensor arguments, "
+                    f"please use None as the respective in_dim"
+                )
+            if in_dim is not None and (in_dim < -arg.dim() or in_dim >= arg.dim()):
+                raise ValueError(
+                    f"vmap({_get_name(func)}, in_dims={in_dims}, ...)(<inputs>): "
+                    f"Got in_dim={in_dim} for some input, but that input is a Tensor "
+                    f"of dimensionality {arg.dim()} so expected in_dim to satisfy "
+                    f"-{arg.dim()} <= in_dim < {arg.dim()}."
+                )
+            if in_dim is not None and in_dim < 0:
+                flat_in_dims[i] = in_dim % arg.dim()
+
+        return (
+            _validate_and_get_batch_size(flat_in_dims, flat_args),
+            flat_in_dims,
+            flat_args,
+            args_spec,
+        )
+
+    functorch._src.vmap._process_batched_inputs = _process_batched_inputs
+
+    def _create_batched_inputs(flat_in_dims, flat_args, vmap_level: int, args_spec):
+        # See NOTE [Ignored _remove_batch_dim, _add_batch_dim]
+        batched_inputs = [
+            arg
+            if in_dim is None
+            else arg.apply(lambda _arg: _add_batch_dim(_arg, in_dim, vmap_level))
+            if isinstance(arg, TensorDictBase)
+            else _add_batch_dim(arg, in_dim, vmap_level)
+            for in_dim, arg in zip(flat_in_dims, flat_args)
+        ]
+        return tree_unflatten(batched_inputs, args_spec)
+
+    functorch._src.vmap._create_batched_inputs = _create_batched_inputs
+
+
+class FunctionalModule(nn.Module):
+    """
+    This is the callable object returned by :func:`make_functional`.
+    """
+
+    def __init__(self, stateless_model):
+        super(FunctionalModule, self).__init__()
+        self.stateless_model = stateless_model
+
+    @staticmethod
+    def _create_from(model, disable_autograd_tracking=False):
+        # TODO: We don't need to copy the model to create a stateless copy
+        model_copy = deepcopy(model)
+        param_tensordict = extract_weights(model_copy)
+        if disable_autograd_tracking:
+            param_tensordict.apply(lambda x: x.requires_grad_(False), inplace=True)
+        return FunctionalModule(model_copy), param_tensordict
+
+    def forward(self, params, *args, **kwargs):
+        # Temporarily load the state back onto self.stateless_model
+        old_state = _swap_state(
+            self.stateless_model, params, return_old_tensordict=_RESET_OLD_TENSORDICT
+        )
+        try:
+            return self.stateless_model(*args, **kwargs)
+        finally:
+            # Remove the loaded state on self.stateless_model
+            if _RESET_OLD_TENSORDICT:
+                _swap_state(self.stateless_model, old_state)
+
+
+class FunctionalModuleWithBuffers(nn.Module):
```

Comment:
So it turns out PyTorch has two ways of using "functional modules". The first is the one that you're aware of, [functorch's make_functional](https://pytorch.org/functorch/stable/generated/functorch.make_functional.html). The second is [torch.nn.utils.stateless.functional_call](https://pytorch.org/docs/stable/generated/torch.nn.utils.stateless.functional_call.html).

As a part of the functorch merge into pytorch we've been tasked with figure out how to deduplicate these APIs. So I'm curious what your thoughts on the two APIs are.

Reply:
It should be feasible to get torch.nn.utils.stateless.functional_call to work with TensorDict as well; we would just modify the API to accept a TensorDict along with dictionaries (maybe it just works today)

Reply:
Seems like functional_call would work with tensordict!
I like that buffers and params are split though: in meta RL you usually optimise either one or the other using dedicated algos. Merging them together is asking the user to split them thereafter which is impractical.

On the functorch side I don't really see the value of getting the params and not the buffers though (FunctionalModule va FunctionalModuleWithBuffers). You can get them and do nothing with it no?

Reply:
Sorry, to clarify, I meant that we have two sets of functional module APIs: there is {functorch.make_functional, functorch.make_functional_with_buffers} and {torch.nn.utils.stateless.functional_call}. We might deprecate one of the two sets. If we deprecate the functorch set, then I'm not quite sure how the torchrl code will look after

Reply:
>  in meta RL you usually optimise either one or the other using dedicated algos

Out of curiosity, what are the algorithms that optimize buffers?

Reply:
Stuff that optimise the gamma and lambda decay, I can find the papers

Reply:
> Sorry, to clarify, I meant that we have two sets of functional module APIs: there is {functorch.make_functional, functorch.make_functional_with_buffers} and {torch.nn.utils. stateless.functional_call}. We might deprecate one of the two sets. If we deprecate the functorch set, then I'm not quite sure how the torchrl code will look after

That's what I understood. I was saying that I like less stateless.functional_call because it does not split params and buffers

Reply:
> >  in meta RL you usually optimise either one or the other using dedicated algos
> 
> 
> 
> Out of curiosity, what are the algorithms that optimize buffers?

Here is one

https://proceedings.neurips.cc/paper/2018/file/2715518c875999308842e3455eda2fe3-Paper.pdf